In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

# Use the kagglehub client library to attach Kaggle resources like competitions, datasets, and models to your session
# Learn more about kagglehub: https://github.com/Kaggle/kagglehub/blob/main/README.md

import kagglehub
# kagglehub.dataset_download('<owner>/<dataset-slug>')

In [2]:
from datasets import load_dataset
import random

ds = load_dataset("iamshnoo/alpaca-cleaned-hindi", split="train")
print(f"Total examples: {len(ds)}")
print(f"Columns: {ds.column_names}")
print(f"\nFirst example:")
print(ds[0])

README.md:   0%|          | 0.00/497 [00:00<?, ?B/s]

data/train-00000-of-00001-9c3cdcdad5ba72(…):   0%|          | 0.00/31.3M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/51760 [00:00<?, ? examples/s]

Total examples: 51760
Columns: ['input', 'instruction', 'output']

First example:
{'input': '', 'instruction': 'स्वस्थ रहने के लिए तीन सुझाव दें।', 'output': '1. संतुलित और पौष्टिक आहार खाएं: अपने भोजन में विभिन्न प्रकार के फल और सब्जियां, दुबला प्रोटीन, पूरे अनाज और स्वस्थ वसा शामिल हों। इससे आपके शरीर को आवश्यक पोषक तत्वों से भरपूर होता है ताकि वह अपने कामकाज को सबसे अच्छा कर सके और इससे पुरानी बीमारियों को रोकने में मदद मिल सकती है। 2. नियमित शारीरिक गतिविधि करें: मजबूत हड्डियों, मांसपेशियों और हृदय स्वास्थ्य को बनाए रखने के लिए व्यायाम महत्वपूर्ण है। हर हफ्ते कम से कम 150 मिनट का मध्यम एरोबिक व्यायाम या 75 मिनट का जोरदार व्यायाम करना लक्ष्य रखें। 3. पर्याप्त नींद लें: पर्याप्त गुणवत्ता की नींद लेना शारीरिक और मानसिक भलाई के लिए महत्वपूर्ण है। यह मनोदशा को विनियमित करने, संज्ञानात्मक कार्य को बेहतर बनाने और स्वस्थ विकास और प्रतिरक्षा कार्य को बढ़ावा देने में मदद करता है। हर रात 7-9 घंटे की नींद लेना लक्ष्य रखें।'}


In [4]:
random.seed(42)
indices = random.sample(range(len(ds)), 30)

for i, idx in enumerate(indices):
    ex = ds[idx]
    print(f"\n{'='*70}")
    print(f"#{i+1} (index {idx})")
    print(f"INSTRUCTION: {ex['instruction']}")
    if ex.get('input', '').strip():
        print(f"INPUT: {ex['input']}")
    print(f"OUTPUT ({len(ex['output'])} chars): {ex['output']}")


#1 (index 41905)
INSTRUCTION: एक संभावित हमले वेक्टर का निर्माण करें जो कमजोरियों का शोषण करता है।
INPUT: सिस्टम SQL इंजेक्शन हमले के लिए कमजोर है।
OUTPUT (952 chars): एक संभावित हमले का वेक्टर जो इस कमजोरता का शोषण कर सकता है, वह निम्नानुसार होगा: 1. हमलावर सिस्टम की वेबसाइट पर एक इनपुट फ़ील्ड की पहचान करता है, जैसे कि एक खोज या लॉगिन फॉर्म, जो SQL इंजेक्शन के लिए संवेदनशील है। 2. हमलावर एक दुर्भावनापूर्ण SQL कथन बनाता है, जिसे अंतर्निहित डेटाबेस से संवेदनशील डेटा को हेरफेर या निकालने के लिए डिज़ाइन किया गया है। उदाहरण के लिए, वे लॉगिन फॉर्म में `' OR 1=1;--` का उपयोगकर्ता नाम दर्ज कर सकते हैं, जो प्रमाणीकरण प्रक्रिया को दरकिनार कर सकता है। 3. सिस्टम इनपुट को संसाधित करता है और इंजेक्ट किया गया SQL कोड निष्पादित किया जाता है, जिससे हमलावर को संवेदनशील जानकारी तक पहुंच मिलती है या संभावित रूप से उन्हें डेटा को दूषित या हटा देने की अनुमति मिलती है। 4. हमलावर इस जानकारी या पहुंच का उपयोग सिस्टम को खतरे में डालने या अपने लाभ के लिए इसका उपयोग करने के लिए कर सकता है। यह हमलावर वेक्टर खराब

* #1 idx=41905  quality=4  flag=harmful_content  notes: SQL injection guide
* #2 idx=7296  quality=2  flag=lazy_output  notes: didn't follow instruction, just rearranged words
* #3 idx=1639  quality=4  notes: good creative task completion, minor gender agreement error.
#4 idx=?  quality=1  flag=identity_output  notes: output identical to input, no task performed
* #5*  idx=18024  quality=2  flag=repetition_loop  notes: good first 4 steps, then collapses into repetition and incoherence1.
* #6 idx=16049  quality=5  notes: clean, follows instruction precisely
* #7 idx=14628  quality=2  flag=degenerate_long_output  notes: items 1-5 good, items 6-10 nonsense, missing items
* #8 (idx 9144) — Score: 3, Flag: ai_refusal. Hindi is natural and grammatically clean, but the model over-refused a harmless creative prompt ("name 3 places in your city") by defaulting to "I am an AI" instead of engaging with the task. This pattern teaches bad behavior for a general-purpose assistant.
* #9 (idx 48265) — Score: 3, Flag: degenerate_long_output.
* #10 idx=6717  quality=2  flag=lazy_output  notes: 4 words instead of 5, list format instead of sentence completion, "heat" as best thing about summer feels off
* #11 idx=44348  quality=3  flag=—  notes: first 60% coherent and historically accurate (Ford, assembly line, mass production); last 40% vague and grammatically tangled; doesn't degenerate into nonsense but tapers off
* #12 idx=48540  quality=4  flag=—  notes: addresses all 3 traits in order with behavioral examples, natural Hindi, appropriate length; slightly formulaic phrasing
*  #13 idx=35741  quality=3  flag=translation_artifact  notes: hashtag output is correct in real-world usage but Hindi instruction wrongly translated "Wonder Woman" to "विचित्र महिला"; Hindi-prompt-to-English-output pattern is bad training signal
* #14 idx=5697  quality=2  flag=english_output  notes: code itself is correct but entire response is English despite Hindi instruction; instruction also has typo "फ़ंक्शटोन" → "फ़ंक्शन"; common pattern for code/technical examples in this dataset
* #15 idx=38698  quality=2  flag=wrong_reasoning  notes: Hindi clean and format correct, but logical ordering is wrong - placed "released" (final event) at start instead of end; correct order is 3→2→4→1→5
* #16 idx=27651  quality=5  flag=—  notes: handled ill-posed question correctly - pointed out height doesn't apply underwater, offered depth info instead; pedagogical reframing done right
* #17 idx=2082  quality=1  flag=repetition_loop  notes: first 60% is coherent creative writing about discovering a winged animal; collapses into repetition of "अद्भुत" then incoherent gibberish; severe enough to be unusable
* #18 idx=1952  quality=5  flag=—  notes: all 3 words used naturally in meaningful sentence about nutrition; proper grammar with subordinating clause; sharp contrast to lazy #2 and #10 on similar task
* #19 idx=6140  quality=5  flag=—  notes: correct 3x3 matrix in row-major order, natural Hindi preamble, appropriate length; clean execution
*  #20 idx=14328  quality=2  flag=wrong_format  notes: not actually a haiku - single prose sentence with commas, no 5-7-5 structure, no line breaks; "क्रिमसन" left as English loanword instead of using Hindi equivalent
* #21 idx=15247  quality=2  flag=broken_instruction  notes: instruction translates both "cause" and "effect" to "प्रभाव", making them identical and the task incoherent; output is internally consistent but conceptually broken because of the instruction; model's reasoning was actually correct but expression is garbled
* #22 idx=33118  quality=2  flag=degenerate_long_output  notes: first 60% solid (evolutionary biology of arachnophobia, natural Hindi); last 40% breaks down into grammatical errors, word repetition ("भाग्य और भाग्य"), circular phrasing, typos ("टिभी"); 5th occurrence of this exact pattern in 22 samples
* #23 idx=39453  quality=5  flag=—  notes: exactly one sentence as requested, natural Hindi with idiomatic phrasing ("क्रांति ला दी है"), substantive content covering two mechanisms (information access, personalization)
* #24 idx=1739  quality=5  flag=—  notes: correctly performed query rewriting — resolved "वहां" to "बर्लिन", added "जर्मनी के" for disambiguation, upgraded "लोग" to formal "जनसंख्या"; meaningful NLP skill executed cleanly
* #25 idx=36781  quality=2  flag=wrong_format  notes: same failure as #20 - not a real haiku, just one prose sentence with commas instead of 5-7-5 three-line structure; all 3 required words present, Hindi clean, but task fundamentally not performed; confirms dataset systematically gets haiku format wrong
* #26 idx=13031  quality=5  flag=—  notes: clean math problem execution — formula stated, values substituted, answer with units; pedagogically structured; natural Hindi math terminology
* #27 idx=46925  quality=3  flag=translation_artifact  notes: conceptually correct heap explanation (max-heap, priority queue, bubble up, sift down) but ASCII tree visualization broken without line breaks; technical terms inconsistently translated ("बुलबुला" for bubble, "सीफ्ट"/"सीफ्टिंग" for sift) where English would be standard
* #28 quality=3  flag=—  notes: generic template critique not specific to the actual poem, invents "consistent meter" that doesn't apply, misquotes opening line (drops "जीवन और प्रकाश का"), final sentence broken with "प्रेरणा" repeated and grammar collapsed; coherent overall but vague and slightly inaccurate
* #29 idx=45962  quality=5  flag=—  notes: correctly removed duplicates while preserving first-occurrence order, natural Hindi preamble, concise output
* #30 idx=35713  quality=2  flag=degenerate_long_output  notes: first 60% well-crafted Memorial Day speech (American context — Gettysburg, Normandy, Vietnam, Iraq, Afghanistan); last 35% collapses with "महानता" repetition loop (4 times in one sentence); 6th occurrence of this exact tail-degeneration pattern in 30 samples; also note: trains American memorial framing rather than Indian (शहीद दिवस, etc.)

## Summary: iamshnoo/alpaca-cleaned-hindi

**Sample:** 30 random examples (seed=42)
**Average quality:** 2.93 / 5
**Verdict:** Salvageable with aggressive filtering — usable as one source among multiple, not as the sole dataset.

### Score distribution
| Score | Count | % |
|-------|-------|---|
| 5 | 8 | 27% |
| 4 | 3 | 10% |
| 3 | 6 | 20% |
| 2 | 11 | 37% |
| 1 | 2 | 7% |

### Failure modes observed (frequency)
| Flag | Count | Description |
|------|-------|-------------|
| degenerate_long_output | 6 | Coherent first ~60%, collapses into repetition/nonsense in tail |
| lazy_output | 2 | Doesn't actually perform the task |
| repetition_loop | 2 | Output stuck repeating same word/phrase |
| wrong_format | 2 | E.g., "haiku" produced as prose with commas |
| translation_artifact | 2 | Proper nouns translated when they shouldn't be |
| harmful_content | 1 | SQL injection guide |
| identity_output | 1 | Output identical to input |
| english_output | 1 | Code task answered in English |
| wrong_reasoning | 1 | Format correct, logic wrong |
| ai_refusal | 1 | Over-refuses harmless creative task |
| broken_instruction | 1 | Instruction itself mistranslated |

### Key patterns

**Length is the strongest quality signal.** All 6 degenerate outputs were >1000 chars. All 8 of the highest-scored (5) examples were <600 chars. The dataset's source model degrades reliably in long outputs.

**Quality is bipolar, not uniform.** 37% of examples score 4-5, 44% score 1-2, only 20% are middling. This is actually good for filtering — bipolar distributions separate cleanly.

**Format-specific tasks are systematically broken.** Both haiku requests (#20, #25) failed identically. Likely all poetry-form requests in this dataset are unreliable.

**Code/technical examples leak English.** When asked technical questions, the source model often defaulted to English output (#14) or partially-translated code terminology (#27).

### Recommended filters for Week 2

1. Drop outputs > 700 chars (catches degeneration — biggest impact)
2. Drop outputs with <30% Devanagari characters (catches English leakage)
3. Drop outputs where `output.strip() == input.strip()` (identity outputs)
4. Drop outputs with 3+ repetitions of any 5+ token n-gram (repetition loops)
5. Drop outputs starting with "मैं एक एआई" or "एक भाषा मॉडल के रूप में" (AI refusal patterns)
6. Drop examples flagged for harmful content (SQL injection, hacking, etc.)

**Estimated retention after filters:** 40-50% of 52K = ~21-26K usable examples.